src/semantics.rs

```diff
@@ -159,6 +174,74 @@ impl Validate for Expression {
     }
 }
 
+#[cfg(feature = "expr_validation")]
+impl GetPredicates for Expression {
+    fn get_predicates(&self) -> Vec<&Predicate> {
+        let mut predicates = Vec::new();
+
+        fn visit<'a, 'b>(expr: &'a Expression, predicates: &mut Vec<&'b Predicate>)
+        where
+            'a: 'b,
+        {
+            match expr {
+                Expression::Logical(l) => match l.as_ref() {
+                    LogicalExpression::And(l, r) => {
+                        visit(l, predicates);
+                        visit(r, predicates);
+                    }
+                    LogicalExpression::Or(l, r) => {
+                        visit(l, predicates);
+                        visit(r, predicates);
+                    }
+                    LogicalExpression::Not(r) => {
+                        visit(r, predicates);
+                    }
+                },
+                Expression::Predicate(p) => {
+                    predicates.push(p);
+                }
+            }
+        }
+
+        visit(self, &mut predicates);
+
+        predicates
+    }
+}
+
+#[cfg(feature = "expr_validation")]
+impl<T> GetFields for T
+where
+    T: GetPredicates,
+{
+    fn get_fields(&self) -> HashSet<String> {
```

Comment:
Are there more forthcoming PRs? I see this function not been used in ffi.rs.
Is it possible that we just define this like `fn get_fields(predicates: &Vec<&Predicate>) -> HashSet<String>`, so that we can reuse the `predicates` list we got from `get_predicates` function, and no need to invoke `self.get_predicates()` again?

Reply:
I don't think `get_fields(predicates: &Vec<&Predicate>)` is a good interface since callers should not really care about what the `Predicate` is. Those two traits are indeed unused yet, but I'll make further changes to it.

Reply:
Done with commit 2cf3480, please check to see if this could be resolved.